os/drivers/testcase/test_kthread_stack_protection.c

```diff
@@ -0,0 +1,170 @@
+ * Pre-processor Definitions
+ ****************************************************************************/
+#define	KTHREAD_STACK_NORMAL_SIZE		300
+#define	KTHREAD_STACK_OVERFLOW_SIZE		4096
+
+/****************************************************************************
+ * Private Declarations
+ ****************************************************************************/
+
+static sem_t test_kthread_stack_sem;
+
+/****************************************************************************
+ * Private Function
+ ****************************************************************************/
+
+#if defined(CONFIG_REG_STACK_OVERFLOW_PROTECTION)
+static void kthread_stack_overflow_array(void)
+{
+	volatile char array_normal[KTHREAD_STACK_OVERFLOW_SIZE];
+
+	array_normal[0] = 'a';
+}
+
+static void kthread_stack_prot_inlimit(void)
+{
+	char array_normal[KTHREAD_STACK_NORMAL_SIZE];
+
+	array_normal[0] = 'a';
+}
+
+static void kthread_stack_overflow_func(int *val)
+{
+	char *ptr;
+	int i = 0;
+	volatile char array_normal[KTHREAD_STACK_NORMAL_SIZE];
+
+	ptr = &array_normal[KTHREAD_STACK_NORMAL_SIZE];
+	for (i = 0; i < KTHREAD_STACK_NORMAL_SIZE; i++) {
+		*ptr = 'a';
+		ptr = ptr - 1;
+	}
+
+	*val = *val + 1;
+	if (*val == (KTHREAD_STACK_OVERFLOW_SIZE / KTHREAD_STACK_NORMAL_SIZE)) {
+		return;
+	}
+	/* recursive call to create stack overflow condition */
+	kthread_stack_overflow_func(val);
+}
+#endif
+
+static int kthread_stack_overflow_prot(int argc, FAR char *argv[])
+{
+	char ch = argv[1][0];
+	char *ptr;
+	int i = 0;
+	char array_normal[KTHREAD_STACK_NORMAL_SIZE];
+
+#if defined(CONFIG_MPU_STACK_OVERFLOW_PROTECTION)
+	if (ch == 'N' || ch == 'n') {
+		dbg("Test Stack access within range\n");
+		ptr = &array_normal[KTHREAD_STACK_NORMAL_SIZE];
+		for (i = 0; i < KTHREAD_STACK_NORMAL_SIZE; i++) {
+			*ptr = 'a';
+			ptr = ptr - 1;
+		}
+		dbg(" Test stack access within range - Passed !!\n");
+	} else if (ch == 'O' || ch == 'o') {
+		dbg("Test Stack access in overflow\n");
+		ptr = &array_normal[KTHREAD_STACK_NORMAL_SIZE];
+		for (i = 0; i < KTHREAD_STACK_OVERFLOW_SIZE; i++) {
+			*ptr = 'b';
+			ptr = ptr - 1;
+		}
+		dbg(" Test stack access in overflow - Failed !!\n");
+	}
+#elif defined(CONFIG_REG_STACK_OVERFLOW_PROTECTION)
+
+	if (ch == 'N' || ch == 'n') {
+		dbg("Test Stack access within range\n");
+		kthread_stack_prot_inlimit();
+		dbg(" Test stack access within range using array size - Passed !!\n");
+	} else if (ch == 'A' || ch == 'a') {
+		dbg("Test Overflow using array declaration\n");
+		kthread_stack_overflow_array();
+		dbg(" Test stack access in overflow - Failed !!\n");
+	} else if (ch == 'F' || ch == 'f') {
+		dbg("Test Overflow using recursive function\n");
+		kthread_stack_overflow_func(&i);
+		dbg(" Test stack access in overflow - Failed !!\n");
+	}
+#endif
+
+	(void)sem_post(&test_kthread_stack_sem);
+
+	return 0;
+}
+
+static int test_kthread_stack_overflow_prot_function(unsigned long arg)
```

Comment:
Let's name the function to represent the codes. This is creating the kernel_thread for stack protection.

Reply:
Have changed name. Please check.